### ISSUE TITLE
[BUG] use dropout in multi-LSTM(GRU)  when GPU index is not '0',  will cause "cublas runtime error" [pytorch0.4.0]

### ISSUE THREAD

**Issue by huguyuehuhu** [pytorch/pytorch#7280]
## Issue description
In  `pytorch0.4.0`,
use dropout in multi-LSTM(GRU)  when GPU index is not '0', will cause ` RuntimeError: cublas runtime error : resource allocation failed at /pytorch/aten/src/THC/THCGeneral.cpp:411` .
for example:
1. if set gpu index as **larger than** 0 (`device = 2`), **with** dropout(`dropout = 0.5`), `num_layers = 2`: cause error;
2. if set gpu index as **larger than** 0 (`device = 2`), **without** dropout(`dropout = 0`), `num_layers = 2`:  cause no error;
3. if set gpu index as  **0** (`device = 0`), **no matter dropout or not**,  `num_layers = 2`: cause no error;
4. `num_layers = 1`: all cases is okay.

In pytorch, because **1 layer LSTM has no dropout**, so all cases is okay. Therefore I guess there is a bug at where  dropout is in LSTM(GRU). but  I can't solve it . Can you do me a help?

## Code example
```
import torch
import torch.nn as nn
import torch.optim as optim
from torch.autograd import Variable

class test_model(nn.Module):
    def __init__(self,
                 in_channel=3,
                 num_class=60,
                 seq_length=10,
                 hidden_size =100,
                 num_layers=2,
                 dropout=0.5
                 ):
        super(test_model, self).__init__()
        self.lstm = nn.LSTM(input_size=in_channel,
                             hidden_size=hidden_size,
                             num_layers=num_layers,
                             batch_first=True,
                             bidirectional=False,
                             dropout=dropout
                            )
        self.fc = nn.Linear(hidden_size*seq_length,num_class)

    def forward(self, x):
        out,_ = self.lstm(x)
        out = out.contiguous().view(out.size(0), -1)
        out = self.fc(out)
        return out

if __name__ == '__main__':
    device = 2
    num_layers = 2 # 1
    dropout = 0.5 # 0

    # cronstruct a dataset
    data = []
    label = []
    epoches = 10
    seq_length = 10
    cuda = True
    for i in range(epoches):
        data.append(torch.randn((2, 10, 3)))
        label.append(torch.randn((2, 60)))

    # model
    model = test_model(num_layers=num_layers,dropout=dropout)
    if cuda:
        model= model.cuda(device=device)
    optimizer = optim.Adam(model.parameters(), lr=0.001)

    # cre = nn.CrossEntropyLoss()
    cre = nn.MSELoss()

    for i in range(epoches):
        if cuda:
            x = Variable(data[i]).cuda(device=device)
            y = Variable(label[i]).cuda(device=device)
        else:
            x = Variable(data[i])
            y = Variable(label[i])

        out_put = model(x)
        loss = cre(out_put, y)
        # print(loss)
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
    print('end')

```

**Comment by zou3519** [comment]
What GPU are you running this on?

**Comment by huguyuehuhu** [comment]
`if __name__ == '__main__':
    device = 2
    num_layers = 2 # 1
    dropout = 0.5 # 0`
I  change the `device`,`dropout` and` layers` in the main function in my code above; I have 4  GPUs, set as  `export CUDA_VISIBLE_DEVICES=0,1,2,3` .

    if set gpu index as larger than 0 (device = 2), with dropout(dropout = 0.5), num_layers = 2: cause error;
    if set gpu index as larger than 0 (device = 2), without dropout(dropout = 0), num_layers = 2: cause no error;
    if set gpu index as 0 (device = 0), no matter dropout or not, num_layers = 2: cause no error;
    num_layers = 1: all cases is okay.

**Comment by zou3519** [comment]
Could I see your `nvidia-smi` output? I'm wondering if you're running out of memory.

**Comment by huguyuehuhu** [comment]
No,it did not out of memory. 




    
    
    



    
        
            
                
                    
                        
                    
                    
                        
                            huguyue2016
                            
                            
                                
                                    邮箱：huguyue2016@ia.ac.cn
                                
                        
                    
                
            
        
    


签名由 网易邮箱大师 定制
On 05/08/2018 22:51, Richard Zou wrote: Could I see your nvidia-smi output? I'm wondering if you're running out of memory.

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/pytorch/pytorch","title":"pytorch/pytorch","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/pytorch/pytorch"}},"updates":{"snippets":[{"icon":"PERSON","message":"@zou3519 in #7280: Could I see your `nvidia-smi` output? I'm wondering if you're running out of memory."}],"action":{"name":"View Issue","url":"https://github.com/pytorch/pytorch/issues/7280#issuecomment-387427720"}}}

**Comment by huguyuehuhu** [comment]
Can you repeat my error through my code?




    
    
    



    
        
            
                
                    
                        
                    
                    
                        
                            huguyue2016
                            
                            
                                
                                    邮箱：huguyue2016@ia.ac.cn
                                
                        
                    
                
            
        
    


签名由 网易邮箱大师 定制
在2018年05月09日 00:12，huguyue2016 写道： 
No,it did not out of memory. 




    
    
    



    
        
            
                
                    
                        
                    
                    
                        
                            huguyue2016
                            
                            
                                
                                    邮箱：huguyue2016@ia.ac.cn
                                
                        
                    
                
            
        
    


签名由 网易邮箱大师 定制
On 05/08/2018 22:51, Richard Zou wrote: Could I see your nvidia-smi output? I'm wondering if you're running out of memory.

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/pytorch/pytorch","title":"pytorch/pytorch","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/pytorch/pytorch"}},"updates":{"snippets":[{"icon":"PERSON","message":"@zou3519 in #7280: Could I see your `nvidia-smi` output? I'm wondering if you're running out of memory."}],"action":{"name":"View Issue","url":"https://github.com/pytorch/pytorch/issues/7280#issuecomment-387427720"}}}

**Comment by zou3519** [comment]
Yes, I am able to reproduce your code on 0.4 and master. We'll take a look; this sounds like a bug.

**Comment by ngimel** [comment]
As a workaround, you can add
```
if cuda:
    torch.cuda.set_device(device)
```
to the beginning of your script.

**Comment by zou3519** [comment]
Hmm it looks like the crash comes earlier. Running with CUDA_LAUNCH_BLOCKING=1 gets me
```
$ CUDA_LAUNCH_BLOCKING=1 python cublas.py
Traceback (most recent call last):
  File "cublas.py", line 71, in <module>
    out_put = model(x)
  File "/home/rzou/pytorch/torch/nn/modules/module.py", line 491, in __call__
    result = self.forward(*input, **kwargs)
  File "cublas.py", line 26, in forward
    out,_ = self.lstm(x)
  File "/home/rzou/pytorch/torch/nn/modules/module.py", line 491, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/rzou/pytorch/torch/nn/modules/rnn.py", line 192, in forward
    output, hidden = func(input, self.all_weights, hx, batch_sizes)
  File "/home/rzou/pytorch/torch/nn/_functions/rnn.py", line 323, in forward
    return func(input, *fargs, **fkwargs)
  File "/home/rzou/pytorch/torch/nn/_functions/rnn.py", line 287, in forward
    dropout_ts)
RuntimeError: CUDNN_STATUS_EXECUTION_FAILED
```
Maybe the inputs are bad, will check those.

**Comment by huguyuehuhu** [comment]
Thank you all very much for your help, it works!

**Comment by zou3519** [comment]
Closed by #7515
